### PR TITLE
8326820: Metadata artificially kept alive

### PR DESCRIPTION
**Pull request by xmas92**
ClassLoaderDataGraph provides APIs for walking different metadata. All the iterators which are not designed to be used by the GC also keep the holder of the CLDs alive and by extensions keeps all metadata alive. This is problematic for concurrent GC as it keeps otherwise unreachable classes from being unloaded and the respective metadata freed. 

This patch changes the default iteration behaviour to not keep the holder alive, with the exception of `loaded_classes_do` (renamed `loaded_classes_do_keepalive`) and `modules_do` (renamed `modules_do_keepalive`) which is used by jvmti APIs that requires that the holder is kept alive.

All other uses consumes all the metadata it queries during its safepoint or before releasing the `ClassLoaderDataGraph_lock`. 

Before this change some jcmd, new jfr chunks and some jfr events, all of which consumed these APIs, could cause class unloading to not occur. 

Been running our internal stress test in an even more stressful mode which without this patch reproduces the metaspace OOME [JDK-8326005](https://bugs.openjdk.org/browse/JDK-8326005) consistently within a few hours. And after this patch it does not.

Currently running tier1-tier8 testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326820](https://bugs.openjdk.org/browse/JDK-8326820): Metadata artificially kept alive (**Bug** - P2)(⚠️ The fixVersion in this issue is [23] but the fixVersion in .jcheck/conf is 24, a new backport will be created when this pr is integrated.)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [0048f8bd](https://git.openjdk.org/jdk/pull/19769/files/0048f8bdd3ea24b9060bf5e37d74bc55beb401c7)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [5f29cfbe](https://git.openjdk.org/jdk/pull/19769/files/5f29cfbe79a741552683ff491360a0c920315747)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [86bb89b0](https://git.openjdk.org/jdk/pull/19769/files/86bb89b02660c07807167af417d604d5e781b653)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19769/head:pull/19769` \
`$ git checkout pull/19769`

Update a local copy of the PR: \
`$ git checkout pull/19769` \
`$ git pull https://git.openjdk.org/jdk.git pull/19769/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19769`

View PR using the GUI difftool: \
`$ git pr show -t 19769`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19769.diff">https://git.openjdk.org/jdk/pull/19769.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19769#issuecomment-2175987738)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiEnvBase.cpp [hunk 1]
```diff
@@ -2339,7 +2339,7 @@ JvmtiModuleClosure::get_all_modules(JvmtiEnv* env, jint* module_count_ptr, jobje
   }
 
   // Iterate over all the modules loaded to the system.
-  ClassLoaderDataGraph::modules_do(&do_module);
+  ClassLoaderDataGraph::modules_do_keepalive(&do_module);
```
Comment (fisk) [inline src/hotspot/share/prims/jvmtiEnvBase.cpp:2342]: Looks like this code exposes an OopHandle backed by the CLD handle area, which isn't a strong root that the GC will start tracing from. So it would seem that we need to keep these oops alive somehow.

File: src/hotspot/share/prims/jvmtiGetLoadedClasses.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ JvmtiGetLoadedClasses::getLoadedClasses(JvmtiEnv *env, jint* classCountPtr, jcla
     // Iterate through all classes in ClassLoaderDataGraph
     // and collect them using the LoadedClassesClosure
     MutexLocker mcld(ClassLoaderDataGraph_lock);
-    ClassLoaderDataGraph::loaded_classes_do(&closure);
+    ClassLoaderDataGraph::loaded_classes_do_keepalive(&closure);
```
Comment (fisk) [inline src/hotspot/share/prims/jvmtiGetLoadedClasses.cpp:108]: This one looks like it might not be safe to exposes without keeping the classes alive.

Comment (fisk) [reply src/hotspot/share/prims/jvmtiGetLoadedClasses.cpp:108]: Looks like we fetch the mirror with Klass::java_mirror() and not Klass::java_mirror_no_keepalive(). Its tempting to think that java_mirror will, unlike its evil twin function, keep the mirror alive. However, this maps to _java_mirror.resolve() vs _java_mirror.peek(). The difference between these was only a thing in single generation ZGC as the _java_mirror is an OopHandle with strong references. Single generation ZGC was the only collector that needed to keep oops alive with strong reference loads - no other collector does that.

In summary, unless you use single generation ZGC, we don't seem to keep the mirrors alive that we expose from here.

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 3]
```diff
@@ -242,21 +242,16 @@ LockedClassesDo::~LockedClassesDo() {
 
 // Iterating over the CLDG needs to be locked because
 // unloading can remove entries concurrently soon.
-template <bool keep_alive = true>
-class ClassLoaderDataGraphIteratorBase : public StackObj {
+class ClassLoaderDataGraph::ClassLoaderDataGraphIterator : public StackObj {
```
Comment (fisk) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:245]: If we now have only a no keepalive variation of this iterator implementation, then perhaps it would be a good idea to have a comment here making it clear that 1) this iterator does not keep the metadata alive and hence that 2) it is up to the user to keep oops alive manually if they are to be exposed in the object graph, or we will crash.

**Review by fisk — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jvmtiGetLoadedClasses.cpp (hunk 2 above)
Comment (xmas92) [reply src/hotspot/share/prims/jvmtiGetLoadedClasses.cpp:108]: `ClassLoaderDataGraph::loaded_classes_do_keepalive` is a `keepalive` iteration. That is it will load the holder of the CLD.
https://github.com/openjdk/jdk/blob/08366b1244775e5892bbbb184660821e8774f37a/src/hotspot/share/classfile/classLoaderDataGraph.cpp#L328-L335

File: src/hotspot/share/prims/jvmtiEnvBase.cpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/prims/jvmtiEnvBase.cpp:2342]: `ClassLoaderDataGraph:: modules_do_keepalive ` is a `keepalive` iteration. That is it will load the holder of the CLD.
https://github.com/openjdk/jdk/blob/08366b1244775e5892bbbb184660821e8774f37a/src/hotspot/share/classfile/classLoaderDataGraph.cpp#L302-L310

**Review by fisk — APPROVED** [review]
Looks good!

**Review by stefank — APPROVED** [review]
Looks good. One question below.

File: src/hotspot/share/classfile/systemDictionary.cpp [hunk 4]
```diff
@@ -1585,7 +1585,7 @@ void SystemDictionary::methods_do(void f(Method*)) {
 
   {
     MutexLocker ml(ClassLoaderDataGraph_lock);
-    ClassLoaderDataGraph::methods_do(f);
+    ClassLoaderDataGraph::methods_do_no_keepalive(f);
```
[the remark this answers is not in the document]
Comment (xmas92) [reply src/hotspot/share/classfile/systemDictionary.cpp:1588]: Probably safer to add the suffix. With a comment. 

`static void classes_do(MetaspaceClosure* it);` Above had a scary name but it has no definition I can find. Seems like it was removed in [JDK-8213346](https://bugs.openjdk.org/browse/JDK-8213346) / 147fc3ed1373286f3a849bf5e8cac83deeb55a3e

**Review by stefank — APPROVED** [review]

**Review by coleenp — CHANGES_REQUESTED** [review]
If the default is to not keep the CLD alive, I don't like that we need the details of the side effect in the name.  Just call it classes_do, etc.  I don't care about no-keepalive in all these callers, if that's the right answer for most of these callers.  Put the side effect in the name of the exceptional cases.

We had the iterator keep things alive for safety in the cases where we could have the CLD unload while we were looking at it.  In all these cases, this wasn't needed?  Still have to work through that.

**Comment by stefank** [comment]
> If the default is to not keep the CLD alive, I don't like that we need the details of the side effect in the name. Just call it classes_do, etc. I don't care about no-keepalive in all these callers, if that's the right answer for most of these callers. Put the side effect in the name of the exceptional cases.

I disagree and probably urged/influenced Axel to name these functions this way. Very few people understand that they need to be careful around liveness of CLDs and klasses when they use these iterators/visitors. We keep seeing bugs in this area. Either because the devs fail to keep the klasses alive, or as in this bug, all klasses become kept alive. We want a name that strongly suggests that these functions are not as innocent as one might think, and that if you are going to use these functions you need to go an look at the comments (or ask someone) what we mean with no-keepalive.

> 
> We had the iterator keep things alive for safety in the cases where we could have the CLD unload while we were looking at it.

It turns out that that safety causes significant problems for seldomly run, concurrent marking GCs. If you repeatedly use any of these APIs you in effect turn off class unloading when running wit ZGC.

> In all these cases, this wasn't needed?

We don't think they are. It would be great if you could take an extra close look at the class redefinition code.

> Still have to work through that.

Thanks!

**Review by coleenp — COMMENTED** [review]
So the problem with the no-keepalive names is that for these functions it's very unclear what exactly you're keeping alive, unless you're thinking about concurrent collection GC, and even then it's a puzzler.  I don't see how it's going to help prevent more bugs.  I don't yet understand the current bug.  The 'resolve' for the CLDG iterator was to temporarily keep that CLD from being unloaded, in the short time that we're iterating on that particular CLD.

The CLDG_lock was added later and it may be that that's what prevents the CLD from being *deleted* while we are holding the lock while iterating.  So that keeps the CLD metadata alive.  The oops in the CLD are not kept alive unless we have resolved the holder which the GC sees as the root.  So if we are accessing oops from the CLD, we need to use the keepalive version.  It's not really keeping the CLD alive.  It's making the oops alive, or accessible until they are handled somewhere by the caller.

So the no-keepalive is not a great name for this, even though it's used when accessing oops in other places.  I think the versions of these functions that access oops that need to be kept alive should have a different name, like modules_do_keeping_oops_alive().  But that's a sentence name.  Maybe modules_do_keepalive() can be read as keeping oops alive.  But classes_do_no_keepalive() is a terrible name because we want the class metadata to be kept alive and the name is disconcerting.

Usually I prefer good names to good comments, but here I think good comments are better.

**Comment by stefank** [comment]
>  The 'resolve' for the CLDG iterator was to temporarily keep that CLD from being unloaded, in the short time that we're iterating on that particular CLD.

This is the crux of the problem. For concurrent GCs, if the 'resolve' is called during a concurrent marking the CLD will be "marked" alive, and will be considered live all the way until we start a new concurrent mark cycle. At that point, we'll try again to figure out if the CLD is dead. If you then again use the iterator during that concurrent marking, the cycle repeats. So, if you tend to use these iterators a lot, we'll never get the chance to unload the classes.

This patch tries to combat that, by changing the iterators. With the patch the iterators hands out objects that are not dead, but they are not considered part of the live object graph. You can use the oops (and its transitive closure) in the CLD as long as you block out safepoints. However, if you try to use them after blocking for a safepoint things will break because the objects are not guaranteed to be a part of the live object graph so nothing kept is keeping them alive. That's what the no-keepalive is intended to refer to.

It would be great if we could figure out a name that hints that the iterators are unsafe to use unless you have understood the above.

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 5]
```diff
@@ -242,21 +242,22 @@ LockedClassesDo::~LockedClassesDo() {
...
 // Iterating over the CLDG needs to be locked because
 // unloading can remove entries concurrently soon.
-template <bool keep_alive = true>
-class ClassLoaderDataGraphIteratorBase : public StackObj {
+// This iterator does not keep the CLD alive.
+// Any CLD OopHandles (modules, mirrors, resolved refs)
+// resolved must be treated as no keepalive. And requires
+// that its CLD's holder is kept alive if they escape the
+// caller's safepoint or ClassLoaderDataGraph_lock
+// critical section.
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:250]: I like this comment.  nit. can you take out the word "soon" since it's nearby.

**Review by coleenp — APPROVED** [review]
I like this change a lot better.  If we run into more places where the oops are escaped from the CLD and need to be kept alive maybe we could try to make this more fool-proof, but it's really subtle.   The places that do not resolve() the holder look good.

**Comment by xmas92** [comment]
Thanks for the reviews.
/integrate
